### PR TITLE
Daxko2 API integration

### PR DESCRIPTION
**Pull request by ygerasimov**
Module allows importing Categories (OpenY's classes) and Offerings (OpenY's sessions) into Drupal.

Depends on migrate_source_csv module. See site openy-daxko.ixm.ca for demo. Under programs you can see "Daxko ..." classes. These are the ones imported from CRM. Example http://openy-daxko.ixm.ca/programs/swimming/swim-lessons/daxko-preschool-stage-2-water-movement/class-times?location=all

I am not putting openy_daxko2 module into installation profile for now. Not sure if we should do that by default.

### CODE REVIEW

**Comment by gundevel** [comment]
Build comment file: 
--------------

 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build270 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build270 |
| Installation process | http://install.openy-dev.ffwua.com/build270/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER_COMPOSER/270/

**Comment by gundevel** [comment]
Acessibility Sniffer: front page checking WCAG2AA http://ci.openymca.org/build270/frontWCAG2AAhtmlcs.txt
Acessibility Sniffer: join page checking WCAG2AA http://ci.openymca.org/build270/joinWCAG2AAhtmlcs.txt
Acessibility Sniffer: locations page checking WCAG2AA http://ci.openymca.org/build270/locationsWCAG2AAhtmlcs.txt
Acessibility Sniffer: schedules page checking WCAG2AA http://ci.openymca.org/build270/schedulesWCAG2AAhtmlcs.txt
Acessibility Sniffer: blog page checking WCAG2AA http://ci.openymca.org/build270/blogWCAG2AAhtmlcs.txt

**Comment by gundevel** [comment]
Behat test results file URL http://ci.openymca.org/build270/profiles/contrib/openy/build/reports/behat and failed screenshots http://ci.openymca.org/build270/profiles/contrib/openy/build/reports/behat/screenshots

**Comment by AlexNetman** [comment]
retest this please

**Comment by gundevel** [comment]
Build comment file: 
--------------

 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build306 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build306 |
| Installation process | http://install.openy-dev.ffwua.com/build306/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER_COMPOSER/306/

**Comment by gundevel** [comment]
Acessibility Sniffer: front page checking WCAG2AA http://ci.openymca.org/build306/frontWCAG2AAhtmlcs.txt
Acessibility Sniffer: join page checking WCAG2AA http://ci.openymca.org/build306/joinWCAG2AAhtmlcs.txt
Acessibility Sniffer: locations page checking WCAG2AA http://ci.openymca.org/build306/locationsWCAG2AAhtmlcs.txt
Acessibility Sniffer: schedules page checking WCAG2AA http://ci.openymca.org/build306/schedulesWCAG2AAhtmlcs.txt
Acessibility Sniffer: blog page checking WCAG2AA http://ci.openymca.org/build306/blogWCAG2AAhtmlcs.txt

**Comment by gundevel** [comment]
Behat test results file URL http://ci.openymca.org/build306/profiles/contrib/openy/build/reports/behat and failed screenshots http://ci.openymca.org/build306/profiles/contrib/openy/build/reports/behat/screenshots

File: modules/custom/openy_daxko2/src/Form/ImportForm.php [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+      fputcsv($fp, $fields);
+    }
+    fclose($fp);
+  }
+
+  public static function batchFinished($success, $results, $operations) {
+    // For some reason imported sessions (offerings) got status 13 instead of 1.
+    // So update them manually or dive into the code and find the bug.
+    db_query('UPDATE {node_field_data} SET status=1 WHERE status=13');
```
Comment (podarok) [inline modules/custom/openy_daxko2/src/Form/ImportForm.php]: Is it possible to use db_update here?
We should think about the possibility to have this code working on SQLite and MSSQL. db_query makes this impossible

File: modules/custom/openy_daxko2/src/Form/ImportForm.php [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        $newRow = [];
+        foreach ($row as $key => $value) {
+          if (in_array($key, ['highlights', 'score', 'type'])) {
+            continue;
+          }
+
+          switch ($key) {
+            case 'start_date':
+            case 'end_date':
+              $value = substr($value, 0, 10);
```
Comment (podarok) [inline modules/custom/openy_daxko2/src/Form/ImportForm.php]: Could you make 10 as a class const to be self-explanatory?

File: modules/custom/openy_daxko2/src/Form/ImportForm.php [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        [
+          'query' => $get,
+          'headers' => [
+            'Authorization' => "Bearer " . $accessToken
+          ],
+        ]);
+
+      $programsResponse = json_decode((string)$response->getBody(), TRUE);
+
+//      drupal_set_message(json_encode($programsResponse));
```
Comment (podarok) [inline modules/custom/openy_daxko2/src/Form/ImportForm.php]: Please, remove debug code

File: modules/custom/openy_daxko2/src/Form/ImportForm.php [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    $base_uri = $config->get('base_uri');
+
+    $accessToken = self::getAccessToken($context, $config);
+
+    $client = new Client(['base_uri' => $base_uri]);
+
+    $after = TRUE;
+    $i = 1;
+
+    unlink('/tmp/programs.csv');
```
Comment (podarok) [inline modules/custom/openy_daxko2/src/Form/ImportForm.php]: This code won't work on 2+ webheads (any enterprise hosting with load balancing). We need to take this into consideration and make this path configurable and use by default Drupal temp folder settings.

File: modules/custom/openy_daxko2/config/install/migrate_plus.migration_group.daxko.yml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+uuid: 8f09303f-c298-4b08-8bb2-512e9c4e6a06
+langcode: en
+status: true
+dependencies: {  }
+id: daxko
+label: Daxko
+description: 'Daxko CRM system'
+source_type: 'Use APIv2 to save categories and offerings into CSV files and then migrate from them'
+module: null
```
Comment (podarok) [inline modules/custom/openy_daxko2/config/install/migrate_plus.migration_group.daxko.yml]: I see an error here. Module can't be null. There is some broken dependency needs to be fixed

**Review by podarok — CHANGES_REQUESTED** [review]
Thanks for the code
It is not perfect, but with small tweaks should be working and a good MVP
Please, fix code review issues and automatic code linters errors

**Comment by AlexNetman** [comment]
I have enabled module for current build.

**Comment by ygerasimov** [comment]
@podarok I've adjusted the code. Meanwhile, I kept CSV files to be saved to /tmp/ folder.

Reasons behind it:

- changing to drupal's temporary directory would make life harder from migration standpoint. The path to file is hardcoded in the config. Please advise if there is a workaround (besides writing own fetcher).
- Regarding load balanced environment. Normally balancers keep opened the session. So once a session is opened user is hitting exactly same webhead. In this way, it is completely safe saving files to /tmp/ folder.

Please advise if we still need to avoid using /tmp/ folder. Then we would need to write custom migration fetcher and point it to drupal's temporary folder.

**Comment by gundevel** [comment]
Build comment file: 
--------------

 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build328 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build328 |
| Installation process | http://install.openy-dev.ffwua.com/build328/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER_COMPOSER/328/

**Comment by gundevel** [comment]
Acessibility Sniffer: front page checking WCAG2AA http://ci.openymca.org/build328/frontWCAG2AAhtmlcs.txt
Acessibility Sniffer: join page checking WCAG2AA http://ci.openymca.org/build328/joinWCAG2AAhtmlcs.txt
Acessibility Sniffer: locations page checking WCAG2AA http://ci.openymca.org/build328/locationsWCAG2AAhtmlcs.txt
Acessibility Sniffer: schedules page checking WCAG2AA http://ci.openymca.org/build328/schedulesWCAG2AAhtmlcs.txt
Acessibility Sniffer: blog page checking WCAG2AA http://ci.openymca.org/build328/blogWCAG2AAhtmlcs.txt

**Comment by gundevel** [comment]
Build comment file: 
--------------

 
| Environment | Link |
| ------------- | ------------- |
| Fresh OpenY installation | http://ci.openymca.org/build329 |
| Upgraded(upgrade path) installation | http://upgrade.openy-dev.ffwua.com/build329 |
| Installation process | http://install.openy-dev.ffwua.com/build329/install.php |

--------------
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.openymca.org:8080/job/PR_BUILDER_COMPOSER/329/

**Comment by gundevel** [comment]
Acessibility Sniffer: front page checking WCAG2AA http://ci.openymca.org/build329/frontWCAG2AAhtmlcs.txt
Acessibility Sniffer: join page checking WCAG2AA http://ci.openymca.org/build329/joinWCAG2AAhtmlcs.txt
Acessibility Sniffer: locations page checking WCAG2AA http://ci.openymca.org/build329/locationsWCAG2AAhtmlcs.txt
Acessibility Sniffer: schedules page checking WCAG2AA http://ci.openymca.org/build329/schedulesWCAG2AAhtmlcs.txt
Acessibility Sniffer: blog page checking WCAG2AA http://ci.openymca.org/build329/blogWCAG2AAhtmlcs.txt

**Comment by gundevel** [comment]
Behat test results file URL http://ci.openymca.org/build328/profiles/contrib/openy/build/reports/behat and failed screenshots http://ci.openymca.org/build328/profiles/contrib/openy/build/reports/behat/screenshots

**Comment by gundevel** [comment]
Behat test results file URL http://ci.openymca.org/build329/profiles/contrib/openy/build/reports/behat and failed screenshots http://ci.openymca.org/build329/profiles/contrib/openy/build/reports/behat/screenshots

**Comment by podarok** [comment]
@ygerasimov 
Could you please record a video about this integration and send avi(any format) to me? I'll post the video to OpenY Youtube channel after release /cc @YMCA-GTC @ymcagtc

**Pull request merged**